src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -245,25 +261,96 @@ private static String extractQualifiedName(DetailAST ast) {
     }
 
     /**
-     * Register to outer super classes of given classAst that
+     * Register to outer super class of given classAst that
      * given classAst is extending them.
      *
-     * @param classAst class which outer super classes will be
+     * @param classAst class which outer super class will be
      *                 informed about nesting subclass
+     * @param qualifiedClassName qualifies class name(with package) of the current class
      */
-    private void registerNestedSubclassToOuterSuperClasses(DetailAST classAst) {
+    private void registerNestedSubclassToOuterSuperClasses(DetailAST classAst,
+                                                           String qualifiedClassName) {
         final String currentAstSuperClassName = getSuperClassName(classAst);
         if (currentAstSuperClassName != null) {
-            for (ClassDesc classDesc : classes) {
-                final String classDescQualifiedName = classDesc.getQualifiedName();
-                if (doesNameInExtendMatchSuperClassName(classDescQualifiedName,
-                        currentAstSuperClassName)) {
+            final List<ClassDesc> sameNamedClassList =
+                    classesWithSameName(currentAstSuperClassName);
+            if (sameNamedClassList.isEmpty()) {
+                final ClassDesc classDesc = innerClasses.get(currentAstSuperClassName);
+                if (classDesc != null) {
                     classDesc.registerNestedSubclass();
                 }
             }
+            else {
+                registerNearestAsSuperClass(qualifiedClassName, sameNamedClassList);
+            }
         }
     }
 
+    /**
+     * Checks if there is a class with same name.
+     *
+     * @param className name of the class
+     * @return true if there is another class with same name.
+     * @noinspection CallToStringConcatCanBeReplacedByOperator
+     */
+    private List<ClassDesc> classesWithSameName(String className) {
```

Comment:
Maybe this connects too much to my other comment, but why not a Set (or TreeSet) ? Can we have multiples that we can't drop?

Reply:
There's no reason to return a  set here. This method returns a list that is iterated once.

Reply:
It is not clear why not, especially since the code was changed to return a Collection which a Set is also subset of.

`Collections.min(Collection<? extends T> coll, Comparator<? super T> comp)`, requires a Comparator in the 2nd argument.
Also using `sorted` method with the stream in `classesWithSameName` also requires a Comparator.

A sorted list will make finding the min the easiest as well since it will always be the first entry if it is not empty.

It is not going to gain us any compare time by combining `classesWithSameName` and `registerNearestAsSuperClass` to just find and return the nearest and keep the collection/list internal? Seems to me this way as it travels through the stream, it will already be tracking for the closest one in addition. `registerNestedSubclassToOuterSuperClasses` just really cares if we find a nearest or not, which in a sorted list should be the first entry.

Reply:
>> A sorted list will make finding the min the easiest as well since it will always be the first entry if it is not empty.

We need just *closest* class. For example,
```
class A {
  class B {}
}

class C {
   class A {
   }
   class E {
     class F extends A {}
   }
 }
```
For class `F`, the closest `A` is `C.A`, not the first class in the list sorted in any way.

Reply:
Done. Your patch has been applied. With one improvement suggested by IDEA:
![image](https://user-images.githubusercontent.com/900805/155855722-576ea2c0-97e2-4ef5-986b-ed65762b5913.png)
this change makes sense, as `sorted` has the complexity `O(N*N)` (in the worst case) while `min` is `O(N)` only.

Reply:
https://stackoverflow.com/questions/49798129/what-is-more-efficient-sorted-stream-or-sorting-a-list
> the sorted() method adds a component to a stream pipeline that captures all of the stream elements into either an array or an ArrayList.
> If an ArrayList is used, you incur the extra overhead of building / growing the list.

>  as the list gets larger, the O(NlogN) sorting will tend to dominate the O(N) overheads of copying. That will mean that the relative difference between the two versions will get smaller.

> Results show that Collections.sort is faster, but not by a big margin